### DISCUSSION TITLE
Is it possible to get the number of required groups in a mixed required/optional regex?

### DISCUSSION THREAD

**Discussion by masklinn** [rust-lang/regex#1204] [Q&A] [answered]
`Regex` has `captures_len()` which returns the total number of groups (required or optional) and `static_captures_len` which returns the number of required groups *if they're all required*.

But in a mixed content situation e.g. `(a)(b)?` the former will return 3 while the latter will return None, I could not find a way to get "2", or to get some sort of bitmap of the required groups. Is there one?

**Comment by BurntSushi** [comment, marked as answer]
There's no out of the box way, no. I'm unsure of the use case for it. But it seems like you can compute it easily enough by looking at `Hir` recursively?

**Reply by masklinn** [reply]
> There's no out of the box way, no. I'm unsure of the use case for it.

I've got a bunch of regexes and replacements which may or may not be templated to include captured data, pretty standard stuff. I wanted to compute the operations to perform up font so the resolution part is simpler e.g. if I know the group is required then I can use indexing and don't need to handle the other case during substitution. That sort of things.

> But it seems like you can compute it easily enough by looking at `Hir` recursively?

Probably, but I figured I might ask in case I'd missed something, or misunderstood `static_captures_len` (which kinda seems to under-deliver).

**Reply by BurntSushi** [reply]
Gotya. Yes, `static_captures_len` under-delivers for your use case, but it is the critical piece that makes `Captures::extract` work in a sensible way.

**Reply by masklinn** [reply]
Out of curiosity, how does it do that? It kinda just seems to be a complicated boolean in that it either returns `None` if any group is optional, or returns `captures_len` which I assume is precomputed?

**Reply by BurntSushi** [reply]
The `Captures::extract` method utilizes const generics. So if the number of capture groups that match can vary based on the match, then what happens when fewer groups match than were requested? It's a footgun that can either be prevented by an immediate panic in all cases where that can _possibly_ happen or by meticulous testing to ensure it doesn't happen on any actual inputs. The latter, I think, would make for a bad API. There are alternative API designs, such as making it fallible, but the whole point of `Captures::extract` is to be a nice ergonomic way to get at capture values. Shoving error handling in there kind of defeats the purpose. It was discussed quite a bit in the [issue requesting the `Captures::extract` feature](https://github.com/rust-lang/regex/issues/824).

The `static_captures_len` is what lets us detect whether the number of capture groups can vary depending on the match. If it can, then `static_captures_len` returns `None` and `Captures::extract` will always panic. And in the case where there are an invariant number of matching capture groups,  if the number of groups requested doesn't match it, then `Captures::extract` will always panic as well. In other words, **if you use `Captures::extract` and it doesn't immediately panic, it's guaranteed to be correct for all possible inputs.**

**Reply by BurntSushi** [reply]
To be clear, `static_captures_len` is a niche routine. But `regex` needs it to make `Captures::extract` work sensibly. So that's why it exists. And it has to be exposed in `regex-syntax` for `regex` to use it.

**Reply by masklinn** [reply]
> The `Captures::extract` method utilizes const generics.

Ah that does clarify things, the documentation for `extract` more clearly exposes the relevant case I had not thought of (despite reading the documentation), which is in the examples of `static_captures_len` but which I skipped over and very much clears the confusion: `(a)|(b)`. In that case `static_captures_len` does return a `Some` *with a different value than `captures_len`*, which is the part that confused me, my testing (and use cases) mostly centered around having optional groups at the end, so I got either `static_captures_len() == None` or `static_captures_len() == Some(captures_len())`, which is why it felt like "a complicated boolean".

This clears things up a lot. Thanks.

> To be clear, `static_captures_len` is a niche routine. But regex needs it to make `Captures::extract` work sensibly. So that's why it exists. And it has to be exposed in regex-syntax for regex to use it.

Right, but it's also exposed by `Regex` itself which is where I got it from.

**Reply by BurntSushi** [reply]
Ahhhh I understand and see the confusion now. Makes sense. Thanks for explaining that!
